tests/unittests/graphOptzTest.cpp

```diff
@@ -872,3 +872,50 @@ TEST_F(GraphOptz, concatElim) {
   // Check that the concat node is gone.
   EXPECT_EQ(countNodeKind(F_, Kinded::Kind::ConcatNodeKind), 0);
 }
+
+// Check the transformation Concat(Reshape(x) * N) -> Reshape(Concat(x * N)).
+TEST_F(GraphOptz, concatReshapes) {
+  const size_t shape[] = {10, 20};
+  const size_t reshape[] = {10, 1, 20};
+  Type t(ElemKind::FloatTy, shape);
+  llvm::SmallVector<NodeValue, 10> inputs;
+  for (size_t i = 0; i < 10; i++) {
+    // Create 10 reshape nodes whose input is variable of shape {10,20}.
+    // The reshape nodes will transform their input from shape of {10,20}
+    // to {10, 1, 20}.
+    Node *var = F_->getParent()->createVariable(ElemKind::FloatTy, shape,
+                                                "input" + std::to_string(i));
+    auto *RN = F_->createReshape("reshape" + std::to_string(i), var, reshape);
+    inputs.push_back(RN);
+  }
+  // Create a Concat node that concatenates the 10 reshape nodes.
+  auto *ConcatN = F_->createConcat("concat", inputs, 1);
+  auto outputShape = ConcatN->getResult().dims();
+  // Need to dereference the RN vector, otherwise the user number of those
+  // nodes would always be positive, making them unable to be removed by DCE.
+  inputs.clear();
+  auto *O = F_->createSave("ret", ConcatN);
+
+  EXPECT_EQ(F_->getNodes().size(), 12);
+
+  ::glow::optimize(F_, CompilationMode::Infer);
+
+  // After optimization, we expect to see only 3 nodes.
+  EXPECT_EQ(F_->getNodes().size(), 3);
+
+  // CN should not exist any more.
+  EXPECT_TRUE(std::find_if(F_->getNodes().begin(), F_->getNodes().end(),
+                           IsSameNodeAddress(ConcatN)) == F_->getNodes().end());
+
+  // The input of O should be a Reshape node now, with the same result shape of
+  // CN.
+  auto *newRN =
+      llvm::dyn_cast<ReshapeNode>(llvm::dyn_cast<SaveNode>(O)->getInput());
+  ASSERT_TRUE(newRN);
+  EXPECT_TRUE(newRN->getType()->dims().equals(outputShape));
+
+  // The input of newRN should be a ConcatNode now.
+  auto *newCN = llvm::dyn_cast<ConcatNode>(
+      llvm::dyn_cast<SaveNode>(O)->getInput()->getNthInput(0));
```

Comment:
why do you need to case O to SaveNode? It's already a SaveNode*.

Reply:
hi, thx for pointing out this. Initially i was following the convention of other tests. But it seems that some cases dyn_cast node O, while some do not. Maybe we should update all the cases.

Reply:
yup, let's do it on a separate PR.